docs/rfc/0003-collaborative-objects.md

```diff
@@ -0,0 +1,649 @@
+the object. Otherwise the change will be merged in to the object and announced
+to the network.
+
+## Implementation
+
+### Automerge
+
+It is useful to briefly outline how automerge functions in order for the
+following to be sensible. Everything automerge does is based on a merging a log
+of operations. An operation might be something like "create a list under the
+'comments' key of the root object", or "insert the character 'a' after the 
+character inserted by the 2nd change actor 1 made". Every operation has an 
+identifier - which allows statements like "the character inserted by the 2nd
+change actor 1 made" to be precise. This operation ID is the combination of a
+unique identifier for each actor, and an always incrementing sequence number.
+This construction, along with sorting by actor IDs in the case of a tie, allows
+us to place operations in a total order which respects causality. i.e if I add
+an operation then no operation that I could have observed at the time I made
+the operation will come after it in the log.
+
+Automerge defines a number of operations along with merge semantics for those
+operations. More detail on that can be found in [the implementation](https://github.com/automerge/automerge)
+and in [the paper](https://arxiv.org/abs/1608.03960).
+
+Operations in automerge are transported in batches called "changes". Each 
+change references zero or more changes it depends on via their hash. In this 
+manner automerge is similar to git in that it's a hash linked graph of changes.
+
+Despite all the complexity under the hood, the API of automerge is relatively
+simple. Automerge works in terms of "documents", a document is a single log of
+changes. Every time you modify an automerge document you generate a new entry 
+for the change log. Each change is just some bytes. When you receive changes 
+from other actors you just pass these changes (which, again, are just bytes) to
+automerge to add to the change log. The end result is that you load a bunch of
+binary changes and get back a JSON object.
+
+There are some subtleties around preserving user intent when modifying
+documents, but these are not too onerous.
+
+### Object Trees
+
+Given that automerge changes are a hash linked graph, we can map them to Git.
+We do so by wrapping each change in a commit. The commit points at a tree with
+the following layout
+
+```
+.
+|--manifest
+|--schema
+|  |--schema.json
+|  |--migrations
+|  |  |--<migration hash>.json
+|  |  |--<migration hash>.json
+|--change
+|--author
+```
+
+This tree contains a single change to a collaborative object. We will go into
+more details shortly. Any direct dependencies of this change are encoded in the 
+same manner and become the parents of this commit. This allows us to
+reconstruct the automerge depdency graph. 
+
+Along with the dependencies of the commit we also need to add the commit of the
+identity which created this commit. We need this identity to validate
```

Comment:
Identity here refers to the head of the identity which is the author of this change. This is the same commit whose hash is referenced in the `author` entry. I've explicitly used the term `identity` because I don't _think_ we need to specify whether it's a person or a project.

Reply:
What would it mean for the `author` to be a `Project`? :thinking:

Reply:
That someone who was a member of the project at that point in time made a change on behalf of the project. I’m not sure how much effort we’d be willing to make to resolve this to “clowntoe@hy... on behalf of sy@hy...“ for rendering, but it makes sense to me semantically =]